File: sdk/ts/test/freestyle.test.ts
import { expect, test } from "bun:test";
import { Code, ConnectError, createRouterTransport } from "@connectrpc/connect";
import { APIError, Client, MeshDriver } from "../src";
import { Freestyle } from "../src/freestyle";
import { SandboxService, SnapshotService, VpcService } from "../src/gen/vmon/v1/api_pb";

interface RecordedRpc {
  method: string;
  input: object;
}

interface FakeState {
  view: unknown;
  /** Per-method response overrides taking precedence over `view`. */
  views: Record<string, unknown>;
  rows: unknown[];
  snapshots: string[];
  exec: { code: bigint; stdout: Uint8Array; stderr: Uint8Array };
  file: Uint8Array;
  missingPaths: Set<string>;
  deniedPaths: Set<string>;
  vpcs: { id: string; name: string; cidr: string; createdAtUnixMillis: bigint }[];
}

/** In-memory vmon gRPC surface exercising the Freestyle facade mappings. */
function fakeVmon() {
  const rpcs: RecordedRpc[] = [];
  const state: FakeState = {
    view: {},
    views: {},
    rows: [],
    snapshots: [],
    exec: { code: 0n, stdout: new Uint8Array(), stderr: new Uint8Array() },
    file: new TextEncoder().encode("payload"),
    missingPaths: new Set(),
    deniedPaths: new Set(),
    vpcs: [],
  };
  const view = (method: string, input: object) => {
    rpcs.push({ method, input });
    return { json: JSON.stringify(state.views[method] ?? state.view) };
  };
  const router = createRouterTransport(({ service }) => {
    service(SandboxService, {
      create: (req) => view("Create", req),
      get: (req) => view("Get", req),
      list: (req) => {
        rpcs.push({ method: "List", input: req });
        return { sandboxesJson: state.rows.map((row) => JSON.stringify(row)) };
      },
      terminate: (req) => view("Terminate", req),
      stop: (req) => view("Stop", req),
      remove: (req) => view("Remove", req),
      suspend: (req) => view("Suspend", req),
      resume: (req) => view("Resume", req),
      resize: (req) => view("Resize", req),
      extend: (req) => view("Extend", req),
      setIdleTimeout: (req) => view("SetIdleTimeout", req),
      snapshot: (req) => view("Snapshot", req),
      execCapture: (req) => {
        rpcs.push({ method: "ExecCapture", input: req });
        return state.exec;
      },
      ptyExec: (req) => {
        rpcs.push({ method: "PtyExec", input: req });
        return state.exec;
      },
      fileRead: (req) => {
        rpcs.push({ method: "FileRead", input: req });
        return { data: state.file };
      },
      fileWrite: (req) => {
        rpcs.push({ method: "FileWrite", input: req });
        return {};
      },
      fileStat: (req) => {
        const path = String(Reflect.get(req, "path"));
        if (state.missingPaths.has(path)) throw new ConnectError("no such path", Code.NotFound);
        if (state.deniedPaths.has(path))
          throw new ConnectError("permission denied", Code.PermissionDenied);
        return view("FileStat", req);
      },
    });
    service(SnapshotService, {
      list: () => ({ snapshots: state.snapshots }),
      restore: (req) => view("Restore", req),
      fork: (req) => view("Fork", req),
    });
    service(VpcService, {
      create: (req) => {
        rpcs.push({ method: "VpcCreate", input: req });
        return {
          id: "vpc-1234abcd",
          name: req.name,
          cidr: req.cidr || "10.77.0.0/16",
          createdAtUnixMillis: 1n,
        };
      },
      list: () => {
        rpcs.push({ method: "VpcList", input: {} });
        return { vpcs: state.vpcs };
      },
      delete: (req) => {
        rpcs.push({ method: "VpcDelete", input: req });
        return {};
      },
    });
  });
  return { rpcs, state, transport: () => router };
}

function connectFreestyle() {
  const fake = fakeVmon();
  const client = new Client(
    new MeshDriver("http://node-a", { discover: false, transport: fake.transport }),
  );
  return { ...fake, freestyle: new Freestyle(client) };
}

function specOf(call: RecordedRpc | undefined): Record<string, unknown> {
  const raw = call === undefined ? undefined : Reflect.get(call.input, "specJson");
  return JSON.parse(String(raw));
}

test("create snapshots default and fully configured requests", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-default" };
  const defaultVm = await freestyle.vms.create();
  expect(defaultVm.vmId).toBe("vm-default");
  expect(specOf(rpcs.at(-1))).toEqual({ timeout_secs: 0, block_network: false });

  state.view = { id: "vm-full" };
  const created = await freestyle.vms.create({
    name: "workspace",
    template: "fs-snap-1",
    cpu: 8,
    memory: 16,
    storage: 80,
    idleTimeoutSeconds: null,
    activityThresholdBytes: 4096,
    persistence: { type: "sticky", priority: 99 },
    nics: [{ vpc: "vpc-1234abcd", mode: "routed", ipv4: "10.88.0.4", default: false }],
    env: { A: "1" },
    workdir: "/workspace",
    tags: { suite: "compat" },
  });
  expect(created.vmId).toBe("vm-full");
  expect(created.vm.vmId).toBe("vm-full");
  expect(created.domains).toEqual([]);
  expect(specOf(rpcs.at(-1))).toEqual({
    name: "workspace",
    template: "fs-snap-1",
    env: { A: "1" },
    workdir: "/workspace",
    tags: { suite: "compat" },
    timeout_secs: 0,
    block_network: false,
    idle_timeout_secs: 0,
    activity_threshold_bytes: 4096,
    persistence: { type: "sticky", priority: 10 },
    nics: [{ vpc: "vpc-1234abcd", ipv4: "10.88.0.4", default: false }],
    cpus: 8,
    memory: 16_384,
    disk_mb: 81_920,
  });
});

test("create rejects multiple NICs before contacting the daemon", async () => {
  const { rpcs, freestyle } = connectFreestyle();
  expect(
    freestyle.vms.create({
      nics: [
        { vpc: "vpc-1234abcd", mode: "routed", ipv4: true },
        { vpc: "vpc-deadbeef", mode: "routed", ipv4: "10.88.0.4" },
      ],
    }),
  ).rejects.toThrow("single NIC");
  expect(rpcs).toEqual([]);
});

test("VPC namespaces map create, list, and delete RPCs", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  const created = await freestyle.vpc.create({ name: "private", cidr: "10.88.0.0/24" });
  expect(created).toEqual({
    vpcId: "vpc-1234abcd",
    vpc: { vpcId: "vpc-1234abcd" },
  });
  expect(rpcs.at(-1)).toMatchObject({
    method: "VpcCreate",
    input: { name: "private", cidr: "10.88.0.0/24" },
  });
  state.vpcs = [
    {
      id: "vpc-1234abcd",
      name: "private",
      cidr: "10.88.0.0/24",
      createdAtUnixMillis: 1n,
    },
  ];
  expect(await freestyle.vpc.list()).toEqual({
    vpcs: [{ vpcId: "vpc-1234abcd", name: "private", cidr: "10.88.0.0/24" }],
  });
  await freestyle.vpc.delete({ vpcId: "vpc-1234abcd" });
  expect(rpcs.at(-1)).toMatchObject({
    method: "VpcDelete",
    input: { id: "vpc-1234abcd" },
  });
});

test("create from a snapshot restores and rejects resizing", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-3" };
  await freestyle.vms.create({
    snapshotId: "snap-1",
    name: "clone",
    idleTimeoutSeconds: 30,
    activityThresholdBytes: 512,
    persistence: { type: "ephemeral" },
  });
  const restore = rpcs.at(-1);
  expect(restore?.method).toBe("Restore");
  expect(Reflect.get(restore?.input ?? {}, "name")).toBe("snap-1");
  expect(JSON.parse(String(Reflect.get(restore?.input ?? {}, "bodyJson")))).toEqual({
    name: "clone",
    timeout_secs: 0,
    block_network: false,
    idle_timeout_secs: 30,
    activity_threshold_bytes: 512,
    persistence: { type: "ephemeral" },
  });
  expect(freestyle.vms.create({ snapshotId: "snap-1", memory: 4 })).rejects.toThrow(
    "fixed by the snapshot",
  );
});

test("exec wraps shell strings and decodes captured output", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-4" };
  const { vm } = await freestyle.vms.create();
  state.exec = {
    code: 3n,
    stdout: new TextEncoder().encode("out"),
    stderr: new TextEncoder().encode("err"),
  };
  const result = await vm.exec("echo out");
  expect(result).toEqual({ stdout: "out", stderr: "err", statusCode: 3 });
  expect(rpcs.at(-1)).toMatchObject({
    method: "ExecCapture",
    input: { id: "vm-4", exec: { cmd: ["sh", "-c", "echo out"] } },
  });
  await vm.exec({ command: "sleep 9", timeoutMs: 1_500 });
  expect(rpcs.at(-1)).toMatchObject({ input: { exec: { timeout: 1.5 } } });
  const terminalResult = await vm.exec({ command: "pwd", terminal: "term-1", timeoutMs: 2_000 });
  expect(terminalResult).toEqual({ stdout: "out", stderr: "err", statusCode: 3 });
  expect(rpcs.at(-1)).toMatchObject({
    method: "PtyExec",
    input: {
      id: "vm-4",
      sessionId: "term-1",
      command: "pwd",
      timeout: 2,
    },
  });
});

test("fs round-trips text and reports existence and stat", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-5" };
  const { vm } = await freestyle.vms.create();
  await vm.fs.writeTextFile("/tmp/hello.txt", "Hello");
  expect(rpcs.at(-1)).toMatchObject({
    method: "FileWrite",
    input: { id: "vm-5", path: "/tmp/hello.txt" },
  });
  state.file = new TextEncoder().encode("Hello");
  expect(await vm.fs.readTextFile("/tmp/hello.txt")).toBe("Hello");

  state.views.FileStat = { type: "dir", size: 4096, mode: 0o40755, mtime: 1_700_000_000 };
  expect(await vm.fs.stat("/tmp")).toEqual({
    size: 4096,
    isFile: false,
    isDirectory: true,
    isSymlink: false,
    permissions: "755",
    modified: new Date(1_700_000_000_000).toISOString(),
  });
  expect(await vm.fs.exists("/tmp")).toBe(true);
  state.missingPaths.add("/nope");
  expect(await vm.fs.exists("/nope")).toBe(false);
  state.deniedPaths.add("/root/secret");
  expect(vm.fs.exists("/root/secret")).rejects.toThrow(APIError);
});

test("stop calls Stop and start after stop calls Resume", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-6" };
  const { vm } = await freestyle.vms.create();
  state.views.Stop = { id: "vm-6", status: "stopped" };
  state.views.Get = { id: "vm-6", observed_state: "stopped" };
  await vm.stop();
  expect(rpcs.map((call) => call.method)).toContain("Stop");
  expect(rpcs.find((call) => call.method === "Stop")).toMatchObject({ input: { id: "vm-6" } });
  state.views.Resume = { id: "vm-6", status: "running" };
  await vm.start();
  expect(rpcs.at(-1)).toMatchObject({ method: "Resume", input: { id: "vm-6" } });
});

test("start sets the idle policy before resuming", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-6-idle" };
  const { vm } = await freestyle.vms.create();
  await vm.start();
  expect(rpcs.at(-1)?.method).toBe("Resume");
  await vm.start({ idleTimeoutSeconds: 900 });
  expect(rpcs.map((call) => call.method).slice(-2)).toEqual(["SetIdleTimeout", "Resume"]);
  expect(Reflect.get(rpcs.at(-2)?.input ?? {}, "idleTimeoutSecs")).toBe(900);
  await vm.start({ idleTimeoutSeconds: null });
  expect(rpcs.map((call) => call.method).slice(-2)).toEqual(["SetIdleTimeout", "Resume"]);
  expect(Reflect.get(rpcs.at(-2)?.input ?? {}, "idleTimeoutSecs")).toBe(0);
});

test("fork snapshots the VM and wraps every clone", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-7" };
  const { vm } = await freestyle.vms.create();
  state.views.Snapshot = { snapshot: "snap-vm-7" };
  state.views.Fork = { clones: [{ id: "fork-a" }, { id: "fork-b" }] };
  const { forks } = await vm.fork({ count: 2 });
  expect(forks.map((fork) => fork.vmId)).toEqual(["fork-a", "fork-b"]);
  expect(forks.map((fork) => fork.vm.vmId)).toEqual(["fork-a", "fork-b"]);
  const fork = rpcs.at(-1);
  expect(fork?.method).toBe("Fork");
  expect(Reflect.get(fork?.input ?? {}, "name")).toBe("snap-vm-7");
  expect(JSON.parse(String(Reflect.get(fork?.input ?? {}, "bodyJson")))).toEqual({ count: 2 });
});

test("vm.snapshot returns the server-assigned snapshot id", async () => {
  const { state, freestyle } = connectFreestyle();
  state.view = { id: "vm-8" };
  const { vm } = await freestyle.vms.create();
  state.views.Snapshot = { snapshot: "snap-8" };
  expect(await vm.snapshot({ name: "base" })).toEqual({
    snapshotId: "snap-8",
    sourceVmId: "vm-8",
  });
});

test("delete removes the sandbox through one RPC", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-9", observed_state: "terminated" };
  await freestyle.vms.delete({ vmId: "vm-9" });
  expect(rpcs.map((call) => call.method)).toEqual(["Remove"]);
});

test("list reports Freestyle states and counts", async () => {
  const { state, freestyle } = connectFreestyle();
  state.rows = [
    {
      id: "a",
      observed_state: "running",
      created_at: 1_000,
      last_active: 2_000,
      last_network_active: 3_000,
    },
    { id: "b", observed_state: "paused" },
    { id: "c", status: "exited" },
  ];
  const listing = await freestyle.vms.list();
  expect(listing.vms.map((row) => [row.id, row.state])).toEqual([
    ["a", "running"],
    ["b", "suspended"],
    ["c", "stopped"],
  ]);
  expect(listing.vms[0]?.createdAt).toBe(new Date(1_000_000).toISOString());
  expect(listing.vms[0]?.lastNetworkActivity).toBe(new Date(3_000_000).toISOString());
  expect(listing.totalCount).toBe(3);
  expect(listing.runningCount).toBe(1);
  expect(listing.suspendedCount).toBe(1);
  expect(listing.stoppedCount).toBe(1);
  expect(listing.startingCount).toBe(0);
});

test("snapshots namespace lists names and rejects unknown ids", async () => {
  const { state, freestyle } = connectFreestyle();
  state.snapshots = ["snap-a", "snap-b"];
  const { snapshots } = await freestyle.vms.snapshots.list();
  expect(snapshots.map((row) => row.snapshotId)).toEqual(["snap-a", "snap-b"]);
  expect(await freestyle.vms.snapshots.get({ snapshotId: "snap-a" })).toMatchObject({
    snapshotId: "snap-a",
    state: "ready",
  });
  expect(freestyle.vms.snapshots.get({ snapshotId: "snap-z" })).rejects.toThrow(APIError);
});

test("resize maps Freestyle units onto ResizeSandboxRequest", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-10" };
  const { vm } = await freestyle.vms.create();
  await vm.resize({ cpu: 8, memory: 16, storage: 80 });
  expect(rpcs.at(-1)).toMatchObject({
    method: "Resize",
    input: { id: "vm-10", cpus: 8, memoryMib: 16_384, diskMb: 81_920n },
  });
});

test("resize rejects non-power-of-two cpu without issuing an RPC", async () => {
  const { rpcs, state, freestyle } = connectFreestyle();
  state.view = { id: "vm-11" };
  const { vm } = await freestyle.vms.create();
  const callsBeforeResize = rpcs.length;
  expect(vm.resize({ cpu: 3 })).rejects.toThrow(TypeError);
  expect(rpcs).toHaveLength(callsBeforeResize);
});

test("constructor maps apiKey/accessToken and fetch onto the vmon driver", async () => {
  const seen: Request[] = [];
  const facade = new Freestyle({
    baseUrl: "http://node-a?discover=off",
    accessToken: "tok-1",
    fetch: (input, init) => {
      seen.push(new Request(input, init));
      return Promise.resolve(Response.json({ ok: true }));
    },
  });
  expect((await facade.client.health()).ok).toBe(true);
  expect(seen[0]?.headers.get("authorization")).toBe("Bearer tok-1");
  await facade.close();
});

File: sdk/go/freestyle/freestyle_test.go
package freestyle

import (
	"context"
	"encoding/json"
	"errors"
	"net"
	"net/http"
	"net/url"
	"reflect"
	"testing"

	vmon "github.com/stencil-hq/vibemon/sdk/go"
	pb "github.com/stencil-hq/vibemon/sdk/go/internal/pb"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

type testDriver struct{ endpoint string }

func (driver *testDriver) Do(context.Context, vmon.DriverRequest) (*http.Response, string, error) {
	return nil, "", errors.New("unexpected HTTP request")
}
func (driver *testDriver) Dial(context.Context, string, url.Values, string) (*vmon.WebSocketConn, string, error) {
	return nil, "", errors.New("unexpected WebSocket request")
}
func (driver *testDriver) Endpoints() []vmon.EndpointInfo {
	return []vmon.EndpointInfo{{URL: driver.endpoint, Healthy: true, Source: "test"}}
}
func (driver *testDriver) Refresh(context.Context, bool) error { return nil }
func (driver *testDriver) Close() error                        { return nil }

type sandboxStub struct {
	pb.UnimplementedSandboxServiceServer
	create         func(context.Context, *pb.CreateSandboxRequest) (*pb.JsonView, error)
	execCapture    func(context.Context, *pb.ExecCaptureRequest) (*pb.ExecCaptureResponse, error)
	setIdleTimeout func(context.Context, *pb.SetIdleTimeoutRequest) (*pb.JsonView, error)
	resume         func(context.Context, *pb.SandboxRef) (*pb.JsonView, error)
	resize         func(context.Context, *pb.ResizeSandboxRequest) (*pb.JsonView, error)
	terminate      func(context.Context, *pb.SandboxRef) (*pb.JsonView, error)
	remove         func(context.Context, *pb.SandboxRef) (*pb.JsonView, error)
	fileStat       func(context.Context, *pb.FilePathRequest) (*pb.JsonView, error)
}

func (stub *sandboxStub) Create(ctx context.Context, request *pb.CreateSandboxRequest) (*pb.JsonView, error) {
	return stub.create(ctx, request)
}
func (stub *sandboxStub) ExecCapture(ctx context.Context, request *pb.ExecCaptureRequest) (*pb.ExecCaptureResponse, error) {
	return stub.execCapture(ctx, request)
}
func (stub *sandboxStub) SetIdleTimeout(ctx context.Context, request *pb.SetIdleTimeoutRequest) (*pb.JsonView, error) {
	return stub.setIdleTimeout(ctx, request)
}
func (stub *sandboxStub) Resume(ctx context.Context, request *pb.SandboxRef) (*pb.JsonView, error) {
	return stub.resume(ctx, request)
}
func (stub *sandboxStub) Resize(ctx context.Context, request *pb.ResizeSandboxRequest) (*pb.JsonView, error) {
	return stub.resize(ctx, request)
}
func (stub *sandboxStub) Terminate(ctx context.Context, request *pb.SandboxRef) (*pb.JsonView, error) {
	return stub.terminate(ctx, request)
}
func (stub *sandboxStub) Remove(ctx context.Context, request *pb.SandboxRef) (*pb.JsonView, error) {
	return stub.remove(ctx, request)
}
func (stub *sandboxStub) FileStat(ctx context.Context, request *pb.FilePathRequest) (*pb.JsonView, error) {
	return stub.fileStat(ctx, request)
}

func testFreestyle(t *testing.T, stub *sandboxStub) *Freestyle {
	t.Helper()
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	server := grpc.NewServer()
	pb.RegisterSandboxServiceServer(server, stub)
	go func() { _ = server.Serve(listener) }()
	t.Cleanup(func() {
		server.Stop()
		_ = listener.Close()
	})
	client := vmon.NewClient(&testDriver{endpoint: "http://" + listener.Addr().String()})
	t.Cleanup(func() { _ = client.Close() })
	return NewFromClient(client)
}

func TestCreateSnapshotsDefaultAndFullyConfiguredSpecs(t *testing.T) {
	var spec map[string]any
	stub := &sandboxStub{}
	stub.create = func(_ context.Context, request *pb.CreateSandboxRequest) (*pb.JsonView, error) {
		if err := json.Unmarshal([]byte(request.GetSpecJson()), &spec); err != nil {
			t.Fatal(err)
		}
		return &pb.JsonView{Json: `{"id":"vm-1","status":"running"}`}, nil
	}
	client := testFreestyle(t, stub)
	ctx := context.Background()
	created, err := client.Vms.Create(ctx, nil)
	if err != nil {
		t.Fatal(err)
	}
	if created.VmID != "vm-1" || created.Vm.VmID != "vm-1" || created.Domains == nil ||
		len(created.Domains) != 0 {
		t.Fatalf("created=%+v", created)
	}
	if want := map[string]any{"timeout_secs": float64(0), "block_network": false}; !reflect.DeepEqual(spec, want) {
		t.Fatalf("default spec=%#v, want %#v", spec, want)
	}

	idle := uint64(0)
	threshold := uint64(4096)
	priority := 99
	useDefault := false
	_, err = client.Vms.Create(ctx, &CreateOptions{
		Name: "mapped", IdleTimeoutSeconds: &idle, ActivityThresholdBytes: &threshold,
		Persistence: &PersistenceOptions{Type: "sticky", Priority: &priority},
		NICs: []NICOptions{
			{VPC: "vpc-1", Mode: "routed", IPv4: "10.88.0.4", Default: &useDefault},
		},
		Template: "fs-snap-1", CPU: 2, Memory: 2, Storage: 3,
		Env: map[string]string{"A": "B"}, Workdir: "/work", Tags: map[string]string{"kind": "test"},
	})
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"name": "mapped", "template": "fs-snap-1", "cpus": float64(2),
		"memory": float64(2048), "disk_mb": float64(3072), "timeout_secs": float64(0),
		"block_network":     false,
		"idle_timeout_secs": float64(0), "activity_threshold_bytes": float64(4096),
		"persistence": map[string]any{"type": "sticky", "priority": float64(10)},
		"nics": []any{
			map[string]any{"vpc": "vpc-1", "ipv4": "10.88.0.4", "default": false},
		},
		"env": map[string]any{"A": "B"}, "workdir": "/work",
		"tags": map[string]any{"kind": "test"},
	}
	if !reflect.DeepEqual(spec, want) {
		t.Fatalf("fully configured spec=%#v, want %#v", spec, want)
	}
}

func TestExecUsesShellAndMapsResult(t *testing.T) {
	var captured *pb.ExecCaptureRequest
	stub := &sandboxStub{}
	stub.execCapture = func(_ context.Context, request *pb.ExecCaptureRequest) (*pb.ExecCaptureResponse, error) {
		captured = request
		return &pb.ExecCaptureResponse{Code: 7, Stdout: []byte("out"), Stderr: []byte("err")}, nil
	}
	client := testFreestyle(t, stub)
	vm, err := client.Vms.Ref(context.Background(), "vm-1")
	if err != nil {
		t.Fatal(err)
	}
	result, err := vm.Exec(context.Background(), ExecOptions{Command: "echo hi", TimeoutMs: 1500})
	if err != nil {
		t.Fatal(err)
	}
	if result != (ExecResult{Stdout: "out", Stderr: "err", StatusCode: 7}) {
		t.Fatalf("result=%+v", result)
	}
	if got := captured.GetExec().GetCmd(); !reflect.DeepEqual(got, []string{"sh", "-c", "echo hi"}) {
		t.Errorf("cmd=%q", got)
	}
	if captured.GetExec().GetTimeout() != 1.5 {
		t.Errorf("timeout=%v", captured.GetExec().GetTimeout())
	}
}

func TestResizeConvertsUnitsAndRejectsBeforeRPC(t *testing.T) {
	calls := 0
	var captured *pb.ResizeSandboxRequest
	stub := &sandboxStub{}
	stub.resize = func(_ context.Context, request *pb.ResizeSandboxRequest) (*pb.JsonView, error) {
		calls++
		captured = request
		return &pb.JsonView{Json: `{"id":"vm-1","status":"running"}`}, nil
	}
	client := testFreestyle(t, stub)
	vm, err := client.Vms.Ref(context.Background(), "vm-1")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := vm.Resize(context.Background(), ResizeOptions{CPU: 3}); err == nil {
		t.Fatal("non-power-of-two CPU was accepted")
	}
	if _, err := vm.Resize(context.Background(), ResizeOptions{Memory: 6}); err == nil {
		t.Fatal("non-power-of-two memory was accepted")
	}
	if calls != 0 {
		t.Fatalf("resize RPC called %d times during client validation", calls)
	}
	if _, err := vm.Resize(context.Background(), ResizeOptions{CPU: 4, Memory: 8, Storage: 12}); err != nil {
		t.Fatal(err)
	}
	if calls != 1 || captured.GetCpus() != 4 || captured.GetMemoryMib() != 8192 || captured.GetDiskMb() != 12288 {
		t.Fatalf("calls=%d request=%+v", calls, captured)
	}
}

func TestStartSetsIdlePolicyBeforeResuming(t *testing.T) {
	var calls []string
	var idleValues []float64
	stub := &sandboxStub{}
	stub.create = func(context.Context, *pb.CreateSandboxRequest) (*pb.JsonView, error) {
		return &pb.JsonView{Json: `{"id":"vm-start"}`}, nil
	}
	stub.setIdleTimeout = func(_ context.Context, request *pb.SetIdleTimeoutRequest) (*pb.JsonView, error) {
		calls = append(calls, "idle")
		idleValues = append(idleValues, request.GetIdleTimeoutSecs())
		return &pb.JsonView{Json: `{"id":"vm-start"}`}, nil
	}
	stub.resume = func(context.Context, *pb.SandboxRef) (*pb.JsonView, error) {
		calls = append(calls, "resume")
		return &pb.JsonView{Json: `{"id":"vm-start"}`}, nil
	}
	client := testFreestyle(t, stub)
	created, err := client.Vms.Create(context.Background(), nil)
	if err != nil {
		t.Fatal(err)
	}
	vm := created.Vm
	if _, err := vm.Start(context.Background(), nil); err != nil {
		t.Fatal(err)
	}
	idle, disabled := uint64(30), uint64(0)
	if _, err := vm.Start(context.Background(), &StartOptions{IdleTimeoutSeconds: &idle}); err != nil {
		t.Fatal(err)
	}
	if _, err := vm.Start(context.Background(), &StartOptions{IdleTimeoutSeconds: &disabled}); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(calls, []string{"resume", "idle", "resume", "idle", "resume"}) {
		t.Fatalf("calls=%v", calls)
	}
	if !reflect.DeepEqual(idleValues, []float64{30, 0}) {
		t.Fatalf("idle values=%v", idleValues)
	}
}

func TestDeleteToleratesNotFound(t *testing.T) {
	removeCalls := 0
	terminateCalls := 0
	stub := &sandboxStub{}
	stub.terminate = func(context.Context, *pb.SandboxRef) (*pb.JsonView, error) {
		terminateCalls++
		return &pb.JsonView{Json: `{}`}, nil
	}
	stub.remove = func(context.Context, *pb.SandboxRef) (*pb.JsonView, error) {
		removeCalls++
		return nil, status.Error(codes.NotFound, "gone")
	}
	client := testFreestyle(t, stub)
	if err := client.Vms.Delete(context.Background(), "vm-gone"); err != nil {
		t.Fatal(err)
	}
	if removeCalls != 1 || terminateCalls != 0 {
		t.Fatalf("remove calls=%d terminate calls=%d", removeCalls, terminateCalls)
	}
}

func TestExistsNarrowsOnlyNotFound(t *testing.T) {
	stub := &sandboxStub{}
	stub.fileStat = func(_ context.Context, request *pb.FilePathRequest) (*pb.JsonView, error) {
		switch request.GetPath() {
		case "/missing":
			return nil, status.Error(codes.NotFound, "missing")
		case "/denied":
			return nil, status.Error(codes.PermissionDenied, "denied")
		default:
			return &pb.JsonView{Json: `{"ok":true,"type":"file","size":3,"mode":420,"mtime":1}`}, nil
		}
	}
	client := testFreestyle(t, stub)
	vm, err := client.Vms.Ref(context.Background(), "vm-1")
	if err != nil {
		t.Fatal(err)
	}
	exists, err := vm.Fs.Exists(context.Background(), "/missing")
	if err != nil || exists {
		t.Fatalf("missing exists=%v err=%v", exists, err)
	}
	exists, err = vm.Fs.Exists(context.Background(), "/present")
	if err != nil || !exists {
		t.Fatalf("present exists=%v err=%v", exists, err)
	}
	if _, err := vm.Fs.Exists(context.Background(), "/denied"); err == nil {
		t.Fatal("permission error was narrowed to absence")
	} else {
		var apiError *vmon.APIError
		if !errors.As(err, &apiError) || apiError.Code != "forbidden" {
			t.Fatalf("permission error=%T %v", err, err)
		}
	}
}
